### ISSUE TITLE
net: sockets: receiving thread not unblocked when closing the blocking socket

### ISSUE THREAD

**Issue by vaussard** [zephyrproject-rtos/zephyr#58280]
**Describe the bug**

I noticed an issue with the condition variable which is used in the socket layer to implement blocking reads. If Thread A calls `recv()` on a blocking socket and Thread B calls `close()` on the same socket, Thread A will keep waiting in `wait_data()` even if the socket is closed.

I did not noticed that issue at first, but this is a problem when reconnecting afterwards. The same socket can be reused but the condition variable is reinitialized in `zsock_socket_internal()` while another thread is still waiting on it.

In order to solve that issue, I signalled the condition variable in `zsock_flush_queue()` to unblock the reader thread, like what is already done in `zsock_received_cb()`:
```
--- a/subsys/net/lib/sockets/sockets.c
+++ b/subsys/net/lib/sockets/sockets.c
@@ -147,6 +147,7 @@ static void zsock_flush_queue(struct net_context *ctx)
 
        /* Some threads might be waiting on recv, cancel the wait */
        k_fifo_cancel_wait(&ctx->recv_q);
+       (void)k_condvar_signal(&ctx->cond.recv);
 }
 
 int zsock_socket_internal(int family, int type, int proto)
```

Is that a proper fix or am I missing something ?

**Expected behavior**

The receiving thread is unblocked when another thread calls `close()` on the blocking socket.

**Comment by rlubos** [comment]
@vaussard Please check https://github.com/zephyrproject-rtos/zephyr/pull/58308
